### PR TITLE
[BUG] Fix BOSS based classifiers truncating class names to single character length

### LINKED ISSUE #4090: [BUG] BOSS classifier class names truncated during prediction
**Describe the bug**
When calling `predict_proba` on IndividualBOSS classifier, sktime returns KeyError

**To Reproduce**

```
import pandas as pd
import numpy as np
x = pd.Series(np.arange(1.0,10.0))
nested_X = pd.DataFrame({
  "nested_X":[x, x, x]
})
y = pd.Series(['01A', '02A', '01A'], dtype="string", name="classes")

from sktime.transformations.panel.padder import PaddingTransformer
from sktime.classification.dictionary_based import IndividualBOSS

iboss = IndividualBOSS()
iboss.fit(nested_X, y)
iboss.predict_proba(nested_X[0:1])
```

** Output behavior **
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[5], line 1
----> 1 iboss.predict_proba(nested_X[0:1])

File c:\Users\anaconda3\envs\env\lib\site-packages\sktime\classification\base.py:262, in BaseClassifier.predict_proba(self, X)
    259     return self._single_class_y_pred(X, method="predict_proba")
    261 # call internal _predict_proba
--> 262 return self._predict_proba(X)

File c:\Users\anaconda3\envs\env\lib\site-packages\sktime\classification\base.py:572, in BaseClassifier._predict_proba(self, X)
    570 dists = np.zeros((n_pred, self.n_classes_))
    571 for i in range(n_pred):
--> 572     dists[i, self._class_dictionary[preds[i]]] = 1
    574 return dists

KeyError: '0'
```

**Expected behavior**
It should return 1 of the 2 unique classes instead of '0'

**Additional context**
I manage to find out that this is due to numpy behavior of changing `str` type to '<U1' which truncates any string to only length of 1.  The particular line of code is at [link](https://github.com/sktime/sktime/blob/8fe509628405eaeaa6cbc26351e946c4a8a54687/sktime/classification/dictionary_based/_boss.py#L619)

Suggest to change it to the following:

```
classes = np.zeros(test_bags.shape[0], dtype=(lambda dtype: "object" if dtype == str else dtype)(type(self._class_vals[0])))
```

Need advise here if I should submit a pull request? Not sure if there's any test cases to run through? Thanks. 

**Versions**
<details>

System:
    python: 3.8.15 (default, Nov 24 2022, 14:38:14) [MSC v.1916 64 bit (AMD64)]
executable: c:\Users\kyunomi\anaconda3\envs\cogpilot\python.exe
   machine: Windows-10-10.0.22621-SP0

Python dependencies:
          pip: 22.3.1
   setuptools: 65.5.0
      sklearn: 1.2.0
       sktime: 0.15.0
  statsmodels: 0.13.5
        numpy: 1.23.5
        scipy: 1.10.0
       pandas: 1.3.5
   matplotlib: 3.5.3
       joblib: 1.1.1
        numba: 0.56.4
     pmdarima: 2.0.1
      tsfresh: 0.20.0

</details>

### PR DESCRIPTION
**Pull request by erjieyong**
#### Reference Issues/PRs

This will fix #4090 which prevent predicted classes list from being populated correctly

#### What does this implement/fix? Explain your changes.

When calling `np.zeros`, change `string` datatype to `object` so that np.zeros will not truncate the string to only the first character.

I've also tested this locally using provided example in the issues as well as running `check_estimator(IndividualBOSS)`. All tests PASSED!

#### Does your contribution introduce a new dependency? If yes, which one?

No

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good to me - could you kindly add a test that certifies for the fix?

I.e., your example (as simple as possible, with dummy data) that fails before the fix and runs after?

It should go into the folder `sktime.classification.dictionary_based`, in a new module `test_boss.py`

Let me know if you don't want to add that and/or need help with `pytest`, best starting point is perhaps following the pattern in `test_tde` (same folder)

**Comment by erjieyong** [comment]
Sure! Thanks for guiding me through this. Happy to learn new things! I will
try to add the unit test and submit again.

**Comment by erjieyong** [comment]
@fkiraly ,I need your help. How do I run the tests after writing them? 

Using `test_tde.py` as an example, when I run the following code: the verbose output does not seem show that the tests in test_tde has been excuted. (eg. test_tde_train_estimate)

```
from sktime.utils.estimator_checks import check_estimator
from sktime.classification.dictionary_based._tde import TemporalDictionaryEnsemble
check_estimator(TemporalDictionaryEnsemble)
```


<details>
  <summary>Output as follows</summary>

```
All tests PASSED!
{'test_clone[TemporalDictionaryEnsemble]': 'PASSED',
 'test_constructor[TemporalDictionaryEnsemble]': 'PASSED',
 'test_create_test_instance[TemporalDictionaryEnsemble]': 'PASSED',
 'test_create_test_instances_and_names[TemporalDictionaryEnsemble]': 'PASSED',
 'test_estimator_tags[TemporalDictionaryEnsemble]': 'PASSED',
 'test_get_params[TemporalDictionaryEnsemble]': 'PASSED',
 'test_has_common_interface[TemporalDictionaryEnsemble]': 'PASSED',
 'test_inheritance[TemporalDictionaryEnsemble]': 'PASSED',
 'test_no_between_test_case_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredict-0]': 'PASSED',
 'test_no_between_test_case_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredict-1]': 'PASSED',
 'test_no_between_test_case_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-0]': 'PASSED',
 'test_no_between_test_case_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-1]': 'PASSED',
 'test_no_cross_test_side_effects_part1[TemporalDictionaryEnsemble]': 'PASSED',
 'test_no_cross_test_side_effects_part2[TemporalDictionaryEnsemble]': 'PASSED',
 'test_repr[TemporalDictionaryEnsemble]': 'PASSED',
 'test_set_params[TemporalDictionaryEnsemble]': 'PASSED',
 'test_set_params_sklearn[TemporalDictionaryEnsemble]': 'PASSED',
 'test_valid_estimator_class_tags[TemporalDictionaryEnsemble]': 'PASSED',
 'test_valid_estimator_tags[TemporalDictionaryEnsemble]': 'PASSED',
 'test_fit_does_not_overwrite_hyper_params[TemporalDictionaryEnsemble-ClassifierFitPredict]': 'PASSED',
 'test_fit_does_not_overwrite_hyper_params[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate]': 'PASSED',
 'test_fit_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_fit_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_fit_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_fit_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_fit_returns_self[TemporalDictionaryEnsemble-ClassifierFitPredict]': 'PASSED',
 'test_fit_returns_self[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate]': 'PASSED',
 'test_fit_updates_state[TemporalDictionaryEnsemble-ClassifierFitPredict]': 'PASSED',
 'test_fit_updates_state[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredict-get_fitted_params]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_methods_have_no_side_effects[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-get_fitted_params]': 'PASSED',
 'test_multiprocessing_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_multiprocessing_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_multiprocessing_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_multiprocessing_idempotent[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredict-get_fitted_params]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_non_state_changing_method_contract[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-get_fitted_params]': 'PASSED',
 'test_persistence_via_pickle[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_persistence_via_pickle[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_persistence_via_pickle[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_persistence_via_pickle[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredict-get_fitted_params]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_raises_not_fitted_error[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-get_fitted_params]': 'PASSED',
 'test_save_estimators_to_file[TemporalDictionaryEnsemble-ClassifierFitPredict-predict]': 'PASSED',
 'test_save_estimators_to_file[TemporalDictionaryEnsemble-ClassifierFitPredict-predict_proba]': 'PASSED',
 'test_save_estimators_to_file[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict]': 'PASSED',
 'test_save_estimators_to_file[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate-predict_proba]': 'PASSED',
 'test_classifier_on_basic_motions[TemporalDictionaryEnsemble]': 'PASSED',
 'test_classifier_on_unit_test_data[TemporalDictionaryEnsemble]': 'PASSED',
 'test_classifier_output[TemporalDictionaryEnsemble-ClassifierFitPredict]': 'PASSED',
 'test_classifier_output[TemporalDictionaryEnsemble-ClassifierFitPredictMultivariate]': 'PASSED',
 'test_handles_single_class[TemporalDictionaryEnsemble]': 'PASSED',
 'test_multivariate_input_exception[TemporalDictionaryEnsemble]': 'PASSED'}
```
</details>

Appreciate your advice. Thanks.

**Comment by fkiraly** [comment]
`check_estimator` only runs the tests that come from `TestAllEstimators` and `TestAllClassifiers` etc.

To run the test in, say, `test_tde` or the new `test_boss`, you need to use `pytest` directly.

Two common options are:
* running the test from your developer IDE, here's the tutorial for visual studio code: https://code.visualstudio.com/docs/python/testing
* running `pytest` via the command line, specifying the module, see here: https://stackoverflow.com/questions/36456920/is-there-a-way-to-specify-which-pytest-tests-to-run-from-a-file

**Comment by erjieyong** [comment]
Thanks. I will try the vscode version. 

I've actually tried running it via command line before, but is facing the following error. Could you enlighten me on what's wrong?

```
pytest sktime\classification\dictionary_based\tests\test_boss.py
```


Error Output

```
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --timeout --cov sktime --cov-report xml --cov-report html sktime\classification\dictionary_based\tests\test_boss.py
  inifile: C:\Users\kyunomi\Desktop\sktime\setup.cfg
  rootdir: C:\Users\kyunomi\Desktop\sktime
```

My pytest version is 7.2.0

**Review by fkiraly — APPROVED** [review]
Thanks, looks great!
Will merge once tests pass.

One tipp, but not a blocker - the tests are a bit repetitive, and they could be made much less copy-paste by using `pytest` `mark.parametrize` (on pairs of the expected `y_pred.dtype`, and the `new_class` dict values).

Feel free to try changing that (you can do it in this PR or a separate one once this is merged), as said it´s not necessary for this to go in imo.

**Comment by fkiraly** [comment]
Ah, but linting is failing now :-)
Here´s the guide: https://www.sktime.org/en/stable/developer_guide/coding_standards.html

**Comment by fkiraly** [comment]
just fixed the linting so we can potentially put this fix in the release

**Comment by erjieyong** [comment]
I see there's another issue. Let me clear it on my end first. I have actually installed Black locally, but I guess it didn't catch the particular portion which failed. Will try and install pre-commit

**Comment by fkiraly** [comment]
> I see there's another issue. Let me clear it on my end first.

Yes, it wants you to write docstrings for the tests. Minimal ones have one line, better ones have an explanation of what precisely the test success and fail conditions are.

**Comment by erjieyong** [comment]
@fkiraly, I've managed to install pre-commit locally already and it has passed all the test. Please review again. Thank you

**Comment by fkiraly** [comment]
linting passes now! 🎉

**Comment by erjieyong** [comment]
Thanks a lot for your guidance. I'll work on improving the tests soon!

**Pull request merged**
